File: challenge2/src/main/java/com/example/demo/Services/ItemService.java
package com.example.demo.Services;

import com.example.demo.Adapter.ItemAdapter;
import com.example.demo.Dtos.ItemDto;
import com.example.demo.Entities.Item;
import com.example.demo.Repository.Interfaces.IDao;
import com.example.demo.Services.Interfaces.IItemService;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Service;

@Service
@Primary
@Qualifier("itemService")
public class ItemService implements IItemService {

  private final IDao<Item> dao;

  @Autowired
  public ItemService(@Qualifier("itemDao") IDao<Item> dao){
    this.dao = dao;  
  }

  @Override
  public List<ItemDto> getItems() {
    return ItemAdapter.toDtos(this.dao.getAll());
  }

  @Override
  public ItemDto getItemById(long id) {
    return ItemAdapter.toDto(this.dao.get(id));    
  }

  @Override
  public ItemDto createItem(ItemDto dto) {
    return ItemAdapter.toDto(this.dao.save(ItemAdapter.create(dto)));    
  }

  @Override
  public ItemDto updateItem(long id, ItemDto dto) {
    var item = this.dao.get(id);
    if(ItemAdapter.update(item, dto)){
      return ItemAdapter.toDto(this.dao.update(item));
    }
    return dto;
  }

  @Override
  public void deleteItem(long id) {
    this.dao.delete(id);
  }
  
  
}

File: py-example/app/service/item_service.py
from typing import List
from fastapi import HTTPException
from ..repository.iitem_repository import IItemRepository
from ..dto.item_dto import ItemCreateDTO, ItemUpdateDTO, ItemDTO
from ..models.item import Item as ItemDB
from .iitem_service import IItemService


class ItemService(IItemService):
    def __init__(self, repository: IItemRepository):
        self.repository = repository
    
    def create_item(self, item_create_dto: ItemCreateDTO) -> ItemDTO:
        db_item = ItemDB(**item_create_dto.dict())
        created_item = self.repository.create(db_item)
        return ItemDTO(**created_item.dict())
    
    def get_item(self, item_id: int) -> ItemDTO:
        db_item = self.repository.get(item_id)
        if db_item is None:
            raise HTTPException(status_code=404, detail="Item not found")
        return ItemDTO(**db_item.dict())
    
    def get_items(self, skip: int = 0, limit: int = 10) -> List[ItemDTO]:
        db_items = self.repository.get_all(skip, limit)
        return [ItemDTO(**db_item.dict()) for db_item in db_items]

    def update_item(self, item_id: int, item_update_dto: ItemUpdateDTO) -> ItemDTO:
        db_item = self.repository.get(item_id)
        if db_item is None:
            raise HTTPException(status_code=404, detail="Item not found")
        updated_item = self.repository.update(item_id, ItemDB(**item_update_dto.dict()))
        return ItemDTO(**updated_item.dict())
    
    def delete_item(self, item_id: int):
        db_item = self.repository.get(item_id)
        if db_item is None:
            raise HTTPException(status_code=404, detail="Item not found")
        self.repository.delete(item_id)